subprojects/language-native/src/integTest/groovy/org/gradle/language/cpp/CppLibraryIntegrationTest.groovy

```diff
@@ -569,4 +569,51 @@ project(':greeter') {
         result.assertTasksExecuted(compileAndLinkTasks(debug), ":assemble")
         sharedLibrary("build/lib/main/debug/hello").assertExists()
     }
+
+    def "can propagate public header generation task through dependencies"() {
+        given:
+        settingsFile << 'include "producer", "consumer"'
+
+        and:
+        file('consumer/src/main/cpp/lib.cpp').text = '#include "common.h"'
+        file('producer/src/main/cpp/lib.cpp').text = '#include "common.h"'
+
+        and:
+        buildFile << '''
+            project(':producer') {
+                apply plugin: 'cpp-library'
+
+                def generatedHeadersDir = file("$buildDir/generated-headers")
+                task generateHeaders {
+                    outputs.dir generatedHeadersDir
+                    doLast {
+                        file("$generatedHeadersDir/common.h").text = '#define FOO'
+                    }
+                }
+
+                library {
+                    publicHeaders.from files(generatedHeadersDir).builtBy(generateHeaders)
+                }
+            }
+
+            project(':consumer') {
+                apply plugin: 'cpp-library'
+
+                tasks.withType(CppCompile) {
+                    doFirst {
+                        assert it.taskDependencies.getDependencies(it)*.path == [":producer:generateHeaders"]
```

Comment:
this isn't the usual pattern we use to verify that the task dependencies are correct. You should instead just verify that the expected tasks were executed, just like we do everywhere else.

Reply:
Very good point. I was so focused on just validating that specific point. I will change it and write the additional tests.